### PR TITLE
Inlet arrays for basic types

### PR DESCRIPTION
**Pull request by aeshapar**
This PR is a feature. It does the following:
- Expands LuaReader's functionalities to process basic Lua arrays 
- Provides Inlet Arrays for basic types (bool, int, double, string)
- Adds Sol as a TPL for Lua and C++ binding

### CODE REVIEW

File: src/axom/inlet/LuaReader.cpp
Comment (white238) [inline src/axom/inlet/LuaReader.cpp:43]: is this warning message correct? what if it is invalid lua?

File: src/axom/inlet/LuaReader.cpp
Comment (white238) [inline src/axom/inlet/LuaReader.cpp:76]: Maybe its time to revisit my getInt getBool getDouble getString functions... @kennyweiss  thoughts?

File: src/axom/inlet/LuaReader.cpp
Comment (white238) [inline src/axom/inlet/LuaReader.cpp:150]: Add a comment clarifying what this if is doing.. "get only numbered items in table" or something.

File: src/axom/inlet/Table.hpp
Comment (white238) [inline src/axom/inlet/Table.hpp:206]: Please document the new parameters.

File: src/axom/inlet/tests/inlet_Inlet.cpp
Comment (white238) [inline src/axom/inlet/tests/inlet_Inlet.cpp:1331]: There needs to be more tests of each function added.  There are several new functions that are not called at all in any tests.  Also an example of how the users would use the new functionality is needed.

File: src/axom/inlet/Table.cpp
Comment (white238) [inline src/axom/inlet/Table.cpp:172]: I'm confused by the new parameters.  Can you explain why they are needed? Are you expecting users to ignore them? If not, maybe move this to an internal function the users wont see.

File: src/axom/inlet/tests/inlet_Inlet.cpp
Comment (aeshapar) [reply src/axom/inlet/tests/inlet_Inlet.cpp:1331]: Agreed on the example. As for the tests, all of the new functions seem to be included in the tests. Could you please point out which function(s) I missed?

Comment (white238) [reply src/axom/inlet/tests/inlet_Inlet.cpp:1331]: Oh i see.. you are doing the addArray's in this test.. then the getMaps above...

File: src/axom/inlet/Table.cpp
Comment (aeshapar) [reply src/axom/inlet/Table.cpp:172]: These are needed for when the addArray functions add the individual elements as fields. I would expect users to ignore them.

Comment (white238) [reply src/axom/inlet/Table.cpp:172]: Then we should move to an internal function that the users dont see.

File: src/axom/inlet/Table.hpp
Comment (white238) [inline src/axom/inlet/Table.hpp:197]: These parameters documentations are wrong for all the get*Array functions

File: src/axom/inlet/examples/arrays.cpp
Comment (white238) [inline src/axom/inlet/examples/arrays.cpp:6]: There is no need for the example function.  Move it to main().  Also return 0 to avoid a warning at the end of main.

Add some comments since this is an example for the users.  For example, "Initialize Inlet" "Parse example input file" "Add verifier that array was found" (also check some of the values inside).

File: src/axom/inlet/Table.cpp
Comment (white238) [inline src/axom/inlet/Table.cpp:114]: pretty sure that '/' is not supposed to be there.

Comment (white238) [reply src/axom/inlet/Table.cpp:114]: Is find specific enough here? is m_name a fully qualified name that may have one embedded in the full path? would endswith (there is axom::utilities::string::endsWith()) be better?

File: src/CMakeLists.txt
Comment (kennyweiss) [inline src/CMakeLists.txt:50]: Is this a strict requirement for inlet / sol ? Or a nice-to-have feature ?

I think this change should be done is a separate PR where it can be tested/validated separately.

Comment (kennyweiss) [reply src/CMakeLists.txt:50]: Also, not all of our users are on `C++14` yet, so we should confirm that this will not be a breaking update.

(I am not expecting problems, but we've been burned in the past, and do not want to unnecessarily introduce problems that will prevent our users from upgrading)

Comment (white238) [reply src/CMakeLists.txt:50]: According to their docs (which they were nice enough to build for us), it should work with C++11 so we can safely revert this change.  This happened early when we weren't sure. 

https://sol2.readthedocs.io/en/v2.20.6/compilation.html

Comment (white238) [reply src/CMakeLists.txt:50]: Though @gzagaris 's PR will soon require C++14 due to raja being updated.  At some point we are going to have to bite that bullet.

File: src/axom/inlet/LuaReader.hpp
Comment (kennyweiss) [inline src/axom/inlet/LuaReader.hpp:18]: Is there a missing edit to `src/axom/inlet/CMakeLists.txt` that adds `sol` as a dependency to `inlet`?
(I realize it is header only, but it should still be added).

File: src/axom/inlet/LuaReader.cpp
Comment (kennyweiss) [reply src/axom/inlet/LuaReader.cpp:76]: Could be -- definitely in a separate PR though.
on the other hand, the current way makes it explicit for the user.

File: src/axom/inlet/LuaReader.cpp
Comment (kennyweiss) [inline src/axom/inlet/LuaReader.cpp:115]: Should these be by-value or by-reference?

File: src/axom/inlet/LuaReader.cpp
Comment (kennyweiss) [inline src/axom/inlet/LuaReader.cpp:132]: Do we always want to extract the data to an explicit map?
Are there potential issues when `T` is not a native type (i.e. int, string, double,bool) ?

**Review by kennyweiss — CHANGES_REQUESTED** [review]
Looks like a nice addition @aeshapar !

I've added some comments.
Mainly, I'm concerned about the change from `C++11` to `C++14` -- this should be tested separately, and possibly deferred to the future if it is not strictly necessary.

Please also:
* document the addition of `sol` in the appropriate places 
* add a note about its license.  @white238 can point out where
* add a line to the RELEASE-NOTES about the new built-in TPL

File: src/axom/inlet/LuaReader.cpp
Comment (white238) [reply src/axom/inlet/LuaReader.cpp:25]: We do have specific changes to the sol.hpp.  its hard to see them in this PR because the file is too long.  but if you view the sol.hpp file and search for axom you can see them there.

File: src/CMakeLists.txt
Comment (aeshapar) [reply src/CMakeLists.txt:50]: C++ 14 was a requirement for sol

Comment (gzagaris) [reply src/CMakeLists.txt:50]: That's correct,  C++14 is required (though no strongly enforced) with newer RAJA. Consequently, everyone is going to be using C++14. All our compilers already support it. We should run with this.

File: src/axom/inlet/tests/inlet_Inlet.cpp
Comment (aeshapar) [reply src/axom/inlet/tests/inlet_Inlet.cpp:1246]: Yes, since sol doesn't distinguish between doubles and ints, they're all just grouped under sol::type::number

File: src/axom/inlet/LuaReader.cpp
Comment (aeshapar) [reply src/axom/inlet/LuaReader.cpp:132]: As of now, we don't provide an interface for extracting non-native types.

File: src/axom/inlet/LuaReader.cpp
Comment (aeshapar) [reply src/axom/inlet/LuaReader.cpp:115]: Since we're repeatedly changing t as we search deeper in the table, by-value

File: src/axom/inlet/CMakeLists.txt
Comment (white238) [inline src/axom/inlet/CMakeLists.txt:44]: Can you change these two lines to just:

```
blt_list_append(TO inlet_depends ELEMENTS sol lua IF SOL_FOUND)
```

File: src/axom/inlet/LuaReader.cpp
Comment (white238) [reply src/axom/inlet/LuaReader.cpp:25]: Can you move this line into sol.hpp with the other defines we had to do?

File: src/axom/inlet/LuaReader.hpp
Comment (white238) [inline src/axom/inlet/LuaReader.hpp:137]: These need to be documented.

File: src/axom/inlet/Reader.hpp
Comment (white238) [inline src/axom/inlet/Reader.hpp:137]: Same documentation here.

File: src/axom/inlet/examples/arrays.cpp
Comment (white238) [inline src/axom/inlet/examples/arrays.cpp:35]: Can you print out the found values here?

File: src/axom/inlet/examples/arrays.cpp
Comment (white238) [inline src/axom/inlet/examples/arrays.cpp:29]: Little strange from a user-perspective to do this in the verify function.  Maybe change this to check that the given values are either "start", "stop", or "pause" to show a semi-real life example.  Then after the verify() call, just loop through the values and print them out like you are now.

**Review by white238 — APPROVED** [review]

File: RELEASE-NOTES.md
Comment (kennyweiss) [inline RELEASE-NOTES.md:59]: Please add: 
* Which version of SOL we are using
* that SOL requires `C++14`
* a link to SOL.

File: src/axom/inlet/LuaReader.hpp
Comment (kennyweiss) [inline src/axom/inlet/LuaReader.hpp:43]: Does this need to be guarded, e.g. with an error message if it cannot open the library, or does SOL/lua already handle that for you?

File: src/axom/inlet/LuaReader.hpp
Comment (kennyweiss) [inline src/axom/inlet/LuaReader.hpp:204]: Suggestion (feel free to ignore):
For readability, I like to separate private methods from private variables with an extra access modifier, e.g.

```cpp
private:
  // Expect this to be called for only Inlet-supported types.
  template<typename T>
  bool getValue(const std::string& id, T& value);

  // Expect this to be called for only Inlet-supported types.
  template <typename T>
  bool getMap(const std::string& id, std::unordered_map<int, T>& values, sol::type type);

private:
  sol::state m_lua;
```

File: src/axom/inlet/Table.cpp
Comment (kennyweiss) [inline src/axom/inlet/Table.cpp:197]: Please run `make style` on your branch before merging.

File: src/axom/inlet/examples/CMakeLists.txt
Comment (kennyweiss) [inline src/axom/inlet/examples/CMakeLists.txt:13]: Should this be guarded on `if(SOL_FOUND)`?  It uses `LuaReader`.

Comment (kennyweiss) [reply src/axom/inlet/examples/CMakeLists.txt:13]: Please build and test this branch against a `C++11`  configuration that has LUA, but not SOL

**Review by kennyweiss — CHANGES_REQUESTED** [review]
Thanks @aeshapar  -- looks like we're almost there.

I had a few suggestions to harden this PR, and to ensure that we don't run into problems related to `C++11` vs. `C++14`.

Please also add a note about SOL and its copyright to the bottom of  `README.md`

If you modified SOL, we should indicate how it was modified. 
@white238  -- perhaps we should also add a patch file?

File: src/axom/inlet/examples/CMakeLists.txt
Comment (white238) [reply src/axom/inlet/examples/CMakeLists.txt:13]: SOL is builtin in so she cant... unless she deletes the header from our repo which seems drastic.

File: src/CMakeLists.txt
Comment (kennyweiss) [reply src/CMakeLists.txt]: Thanks @gzagaris.  Per our team meeting on 9/14, we will hold off on requiring `C++14` for axom for a few months, but can certainly optionally build with `C++14`

File: src/axom/inlet/examples/CMakeLists.txt
Comment (kennyweiss) [reply src/axom/inlet/examples/CMakeLists.txt:13]: Got it, so  `SOL_FOUND` and `LUA_FOUND` should always have the same value?

File: src/axom/inlet/LuaReader.hpp
Comment (aeshapar) [reply src/axom/inlet/LuaReader.hpp:43]: It's not needed for our usage of Lua, so I've removed it.

File: src/axom/inlet/examples/CMakeLists.txt
Comment (white238) [reply src/axom/inlet/examples/CMakeLists.txt]: Not necessarily... for example if you give Axom a LUA_DIR, LUA_FOUND will be true but if you are C++11, SOL_FOUND will be false.  I guarded all the places that were causing problems though.

File: src/axom/inlet/LuaReader.hpp
Comment (white238) [reply src/axom/inlet/LuaReader.hpp:205]: @aeshapar can you do this or just put a line break between the member functions and variables?

**Review by white238 — APPROVED** [review]

File: RELEASE-NOTES.md
Comment (kennyweiss) [inline RELEASE-NOTES.md:60]: 👍

File: src/thirdparty/sol/sol.patch
Comment (kennyweiss) [inline src/thirdparty/sol/sol.patch:2]: 👍

**Review by kennyweiss — COMMENTED** [review]
Thanks @aeshapar -- looks great!

**Review by kennyweiss — APPROVED** [review]

**Pull request merged**
